packages/space-opera/src/components/materials_panel/materials_panel.ts

```diff
@@ -121,11 +139,54 @@ export class MaterialPanel extends ConnectedLitElement {
   }
 
   getMaterial() {
-    return getModelViewer()!.model!.materials[this.selectedMaterialIndex];
+    const materials = getModelViewer()!.model!.materials;
+    return materials[this.selectedMaterialIndex];
+  }
+
+  getMaterialVariant() {
+    let material = this.getMaterial();
+
+    // Creates a new material instance if it does not currently exist under the
+    // variant.
+    if (this.selectedVariant != null &&
+        this.variantSelector.selectedIndex !== 0 &&
+        !material.hasVariant(this.selectedVariant)) {
+      // Creates unique material instance for this variant if one does not
+      // exist.
+      const clone = getModelViewer().model!.createVariantFromMaterial(
+          this.selectedMaterialIndex,
+          material.name + ' ' + this.selectedVariant,
+          this.selectedVariant,
+          true)!;
+
+      const sourceIndex = this.originalMaterialFromCloneMap.get(material.index);
+      if (sourceIndex != null) {
+        this.originalMaterialFromCloneMap.set(clone.index, sourceIndex);
+      } else {
+        this.originalMaterialFromCloneMap.set(clone.index, material.index);
+      }
+
+      this.updateSelectableMaterials();
+      // Cloned material becomes the selected material.
+      this.materialSelector.selectedIndex =
+          this.selectableMaterials.indexOf(clone);
+      if (this.materialSelector.selectedIndex === -1) {
+        console.error('Could not select the new variant material');
+      }
+      this.requestUpdate();
+      material = clone;
+    }
+
+    return material;
   }
 
   getOriginalMaterial() {
-    return this.originalGltf!.materials[this.selectedMaterialIndex];
+    if (this.originalGltf!.materials.length > this.selectedMaterialIndex) {
+      return this.originalGltf!.materials[this.selectedMaterialIndex];
+    } else {
+      return this.originalGltf!.materials[this.originalMaterialFromCloneMap.get(
+          this.selectedMaterialIndex)!];
```

Comment:
nit: this could probably just be a ternary op on the  index.

Reply:
Done